cirq-core/cirq/work/observable_measurement.py

```diff
@@ -354,6 +355,47 @@ def _to_sweep(param_tuples):
     return to_sweep
 
 
+def _parse_checkpoint_options(
+    checkpoint: bool, checkpoint_fn: Optional[str], checkpoint_other_fn: Optional[str]
+) -> Tuple[Optional[str], Optional[str]]:
+    """Parse the checkpoint-oriented options in `measure_grouped_settings`.
+
+    This function contains the validation and defaults logic. Please see
+    `measure_grouped_settings` for documentation on these args.
+
+    Returns:
+        checkpoint_fn, checkpoint_other_fn: Parsed or default filenames for primary and previous
+            checkpoint files.
+    """
+    if not checkpoint:
+        if checkpoint_fn is not None or checkpoint_other_fn is not None:
+            raise ValueError(
+                "Checkpoint filenames were provided but `checkpoint` was set to False."
+            )
+        return None, None
+
+    if checkpoint_fn is None:
+        checkpoint_dir = tempfile.mkdtemp()
+        chk_basename = 'observables'
+        checkpoint_fn = f'{checkpoint_dir}/{chk_basename}.json'
+
+    if checkpoint_other_fn is None:
+        checkpoint_dir = os.path.dirname(checkpoint_fn)
+        chk_basename = os.path.basename(checkpoint_fn)
+        chk_basename, _, ext = chk_basename.rpartition('.')
+        if ext != 'json':
+            raise ValueError(
+                "Please use a `.json` filename or fully "
+                "specify checkpoint_fn and checkpoint_other_fn"
+            )
+        if checkpoint_dir == '':
+            checkpoint_other_fn = f'{chk_basename}.prev.json'
+        else:
+            checkpoint_other_fn = f'{checkpoint_dir}/{chk_basename}.prev.json'
+
+    return checkpoint_fn, checkpoint_other_fn
```

Comment:
Good point. As written, it will still "work" as expected if you use the same filename for both. You just may get corruption if the process dies during the checkpointing process. Let me think about this.

1. it's confusing and the user may be making a mistake
2. a power user may not want two different files around and specifically wants to roll the dice with non-atomic checkpointing

Reply:
I added a test that uses the same name for both. Let me know if you think I should disallow this behavior. I don't have a strong opinion either way

Reply:
Is case (2) expected to be common? e.g. are these files particularly large, or are users expected to create many of them at once? Otherwise, I think the risk of confusion from (1) outweighs the utility of this behavior.

If case (2) _is_ common, it may still be useful to guard it with an "are you sure?"-type flag, similar to the `permit_terminal_measurements` guard on `simulate_expectation_values`:
https://github.com/quantumlib/Cirq/blob/fb43b84c2748c128284491e6b2e65fa605af5a23/cirq-core/cirq/sim/sparse_simulator.py#L223-L228

Reply:
I banned it